### PR TITLE
[#3] Implement basic TOML parser

### PR DESCRIPTION
**Pull request by ChShersh**
This PR introduces basic implementation of TOML parser. Also, some types in `Toml.Type` module were refactored. This resulted in a lot of changes... Though, it works! You can checkout parseable example in `test.toml` file.

### CODE REVIEW

**Comment by ChShersh** [comment]
Argh, `-XDerivingStrategies` doesn't work on GHC-8.0.2... I guess I need to drop them in that case...

File: src/Toml/Parser.hs [hunk 1]
```diff
@@ -1,13 +1,184 @@
...
+-}
+groupByPrefix :: [TableHeader] -> [NonEmpty TableHeader]
+groupByPrefix []     = []
+groupByPrefix (x:xs) = let (y, ys) = span (isThPref x) xs in (x :| y) : groupByPrefix ys
+  where
+    isThPref :: TableHeader -> TableHeader -> Bool
+    isThPref = isPrefixOf `on` unKey . thName
+
+isPrefixOf :: Eq a => NonEmpty a -> NonEmpty a -> Bool
+(x :| xs) `isPrefixOf` (y :| ys) = x == y && List.isPrefixOf xs ys
```
[the remark this answers is not in the document]
Comment (vrom911) [reply src/Toml/Parser.hs]: Oh, I see it's not, it works with simple list as prefix :thinking:

File: test.toml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+a = true
+list = ["one", "two"]
+
+[table.name.2]
+  2Inner = +42
+
+[table.name.1]
+  listInner."google.com" = [true, false]
```
Comment (vrom911) [inline test.toml]: Sorry, I've seen this red highlight in `toml` repo also, why is that?

File: tomland.cabal [hunk 3]
```diff
@@ -26,13 +26,16 @@ library
 
   build-depends:       base >= 4.9 && < 5
                      , hashable
+                     , megaparsec
+                     , parser-combinators
```
Comment (vrom911) [inline tomland.cabal]: Why is this needed?

File: src/Toml/Parser.hs [hunk 4]
```diff
@@ -1,13 +1,184 @@
...
+-- TOML parser
+----------------------------------------------------------------------------
+
+-- Keys
+
+bareKeyP :: Parser Text
+bareKeyP = lexeme $ Text.pack <$> bareStrP
+  where
+    bareStrP :: Parser String
+    bareStrP = some $ alphaNumChar <|> char '_' <|> char '-'
```
[the remark this answers is not in the document]
Comment (ChShersh) [reply src/Toml/Parser.hs]: Yeah, restricting to something like _`-` can't be the first character_ will complicate things a lot... Though, `-_-_-_-` is a valid key in TOML, I hope people won't use such keys...

File: src/Toml/Parser.hs [hunk 5]
```diff
@@ -1,13 +1,184 @@
...
+tableNameP = lexeme $ between (char '[') (char ']') keyP
+
+-- Values
+
+boolP :: Parser Bool
+boolP = False <$ text "false"
+    <|> True  <$ text "true"
+
+-- dateTimeP :: Parser DateTime
+-- dateTimeP = error "Not implemented!"
```
[the remark this answers is not in the document]
Comment (ChShersh) [reply src/Toml/Parser.hs]: Yeah, Will create separate issue for this...

File: src/Toml/Parser.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ChShersh) [reply src/Toml/Parser.hs]: @vrom911 I was also surprised by this behavior. There's:

```haskell
isPrefixOf :: [a] -> [a] -> Bool
isPrefixOf :: [a] -> NonEmpty a -> Bool
```

But somehow there's no

```haskell
isPrefixOf :: NonEmpty a -> NonEmpty a -> Bool
```
:thinking: :man_shrugging:

File: tomland.cabal (hunk 3 above)
Comment (ChShersh) [reply tomland.cabal]: `parser-combinators` is a dependency for `megaparsec`. Some parser combinators are not exported by `megaparsec`. Instead more general versions of them are in the `parser-combinators` packages. Specifically, I need these ones:

```haskell
sepBy1   :: MonadPlus m => m a -> m sep -> m (NonEmpty a)
between  :: Applicative m => m open -> m close -> m a -> m a
sepBy    :: Alternative m => m a -> m sep -> m [a]
manyTill :: Alternative m => m a -> m end -> m [a]
```

File: src/Toml/Parser.hs [hunk 6]
```diff
@@ -1,13 +1,184 @@
...
+-- Keys
+
+bareKeyP :: Parser Text
+bareKeyP = lexeme $ Text.pack <$> bareStrP
+  where
+    bareStrP :: Parser String
+    bareStrP = some $ alphaNumChar <|> char '_' <|> char '-'
+
+stringP :: Parser Text
+stringP = lexeme $ Text.pack <$> (char '"' *> anyChar `manyTill` char '"')
```
[the remark this answers is not in the document]
Comment (ChShersh) [reply src/Toml/Parser.hs]: Well, I'm not sure how to write such parser properly... Because `'` is for literal strings. With `'` I can write:

```toml
quoted   = 'Tom "Dubs" Preston-Werner'
```

But with `"` this needed to be write like this:

```toml
quoted   = "Tom \"Dubs\" Preston-Werner"
```

Actually, I'm not sure which behavior is currently implemented and what be the result of this parser... In `""` strings I also need to support this:

```
\b         - backspace       (U+0008)
\t         - tab             (U+0009)
\n         - linefeed        (U+000A)
\f         - form feed       (U+000C)
\r         - carriage return (U+000D)
\"         - quote           (U+0022)
\\         - backslash       (U+005C)
\uXXXX     - unicode         (U+XXXX)
\UXXXXXXXX - unicode         (U+XXXXXXXX)
```

Not sure how to do this easilty and what comes out of the box...

[the remark this answers is not in the document]
Comment (ChShersh) [reply src/Toml/Parser.hs]: Okay, `"Tom \"Dubs\" Preston-Werner"` is not even parseable because current parser interprets `\` as a separate character which makes sense. So I just need to replace `"` with `'` to have parser of literal strings.

File: src/Toml/Type.hs [hunk 7]
```diff
@@ -122,7 +120,9 @@ arr6 = [ 1, 2.0 ] # INVALID
     -}
     Array  :: [Value t] -> Value 'TArray
 
--- | Untyped 'Value'.
+-- TODO: move into Toml.Type.Internal module then?..
```
Comment (vrom911) [inline src/Toml/Type.hs]: That's good point, because it's exported at the moment..

File: src/Toml/Type.hs [hunk 8]
```diff
@@ -120,7 +120,9 @@ arr6 = [ 1, 2.0 ] # INVALID
     -}
     Array  :: [Value t] -> Value 'TArray
 
--- | Untyped 'Value'.
+-- TODO: move into Toml.Type.Internal module then?.. But it uses 'DateTime' which is not internal...
```
Comment (vrom911) [inline src/Toml/Type.hs]: :disappointed:
